composer/utils/data.py

```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
+                          transform: Callable,
+                          location: Union[str, type] = "end",
+                          pre_post: str = "pre") -> Dataset:
+    """Flexibly add a transform to the dataset's collection of transforms. Warning: this
+    function will not behave as expected if a dataset's collection of transforms contains
+    nested torchvision.transforms.Compose (e.g dataset.transforms =
+    transforms.Compose([transform0, transform1, transforms.Compose(...)])). This function
+    should be updated to flatten nested compositions.
 
     Args:
         dataset: A torchvision-like dataset
         transform: Function to be added to the dataset's collection of transforms
+        location (str, type), optional: Where to insert the transform in
+            the sequence of transforms. "end" will append to the end or "start" will
+            insert at index 0, a transform (e.g. torchvision.transforms.ToTensor) will
+            insert pre- or post- the first instance of `location`, determined by the
+            `pre_post` argument. Default = "end". If `location` is not in list of
+            transforms, will append to end.
+        pre_post (str), optional: Whether to insert before ("pre") or after ("post") the
+            transform string passed in 'location'. Ignored if transform = "start" or "end". Default = "pre".
 
     Returns:
         The original dataset. The transform is added in-place.
     """
 
     if not hasattr(dataset, "transform"):
-        raise ValueError(f"Dataset of type {type(dataset)} has no attribute 'transform'. Expected TorchVision dataset.")
+        raise AttributeError(textwrap.dedent(f"""Dataset must have attribute `transform`."""))
+
+    valid_non_transform_locations = ["start", "end"]
+    valid_pre_post_values = ["pre", "post"]
+    if location not in valid_non_transform_locations and pre_post not in valid_pre_post_values:
+        raise ValueError(
+            textwrap.dedent(
+                f"Invalid value combination for argument `pre_post`: `{pre_post} and `location`: {location}. If location is not one of ['start', 'end'], `pre_post` must be one of ['pre', 'post']."
+            ))
+    if not hasattr(dataset, "transform"):
+        raise AttributeError(textwrap.dedent(f"Argument `dataset` must have attribute `transform`."))
+
+    added_transform = 0
+
+    def added_transform_warning():
+        log.warning(
+            f"Transform {transform} added to dataset. Dataset now has the following transforms: {dataset.transform}")
+        return 1
 
     if dataset.transform is None:
         dataset.transform = transform
-    elif hasattr(dataset.transform, "transforms"):  # transform is a Compose
-        dataset.transform.transforms.append(transform)
-    else:  # transform is some other basic transform, join using Compose
-        dataset.transform = transforms.Compose([dataset.transform, transform])
-
+        added_transform = added_transform_warning()
```

Comment:
For readabiity, can you remove the helper function and split it into two lines? Makes it more readable what's happening.

```suggestion
        log.warn(LOG_MESSAGE)
        added_transform = True
```
Same thing in the other places...